### ISSUE TITLE
0.0.267 has been retagged

### ISSUE THREAD

**Issue by dvzrv** [charliermarsh/ruff#4468]
Hi! I'm one of the package maintainers for ruff on Arch Linux (cc @inglor, @alerque).
We noticed (see downstream ticket: https://bugs.archlinux.org/task/78534), that [0.0.267](https://github.com/charliermarsh/ruff/releases/tag/v0.0.267) has been retagged (i.e. the tag has been deleted and added again, on a different commit).

I would like to point out, that this is problematic (for us and in fact all consumers of this project):

* It may mean malicious activitiy on a project: An attacker may have acquired credentials of one of the developers and changed a release, adding potentially harmful functionality ([supply chain attacks](https://en.wikipedia.org/wiki/Supply_chain_attack) are a thing). This scenario requires all downstreams to manually verify what is going on upstream and write a ticket (e.g. this one).
  As this project does not rely on signed tags, that can be traced back to a separately verifiable entity (btw, the github signing service is pointless to use for tags, as all developers have access to it) and the newly added tag is not signed either, this raises suspicion.
* It breaks all downstream builds relying on the initial tarball, as it will no longer be available. This by proxy breaks [reproducible builds](https://reproducible-builds.org/), which Arch Linux works on: https://reproducible.archlinux.org/.
  In effect this also breaks any derivate distribution of Arch Linux (or other distribution), that relies on their parent distribution's package sources (in our case e.g. Arch Linux 32, Arch Linux ARM, etc.)

TL;DR: Please never retag, just create a new tag.

**Comment by alerque** [comment]
I have the original tar file used to build the Arch package. There are several differences but they all look safe enough to me. They added an svg file to the repo, used it to add branding to the readme, re-indented some YAML, and changed the release workflow from a release dispatch trigger to a tag push trigger. Definitely not a good reason for re-tagging, but not malicious.

As David said, please don't retag after a release goes out. If something goes wrong in the release cycle just increment the tag and try again. That is much less workload on downstream packagers to deal with.

No time is "short enough" either, many of us get instant release notifications and can have a tarball downloaded for packaging in seconds from the tag being posted. If it later gets retagged we may not even be aware of the change until things start going wrong further downstream. Worst case scenario if something is actively harmful in a release yank the tag, but don't repost the same one, increment it again in the next release and explain in release notes what happened to the missing release.

**Comment by alerque** [comment]
We already resolved this specific instance by manually diffing tarballs and re-publishing a new build (incrementing our release number of course).

I would suggest a different approach to the GH workflows. You already need to build everything on every commit anyway to test. Rather than trying to have a different workflow for releases, just add a step to the build workflow that checks whether the commit is tagged and if so, uploads the artifacts to the tag. Here is an example of a workflow that does just that: https://github.com/sile-typesetter/sile/blob/master/.github/workflows/build.yml#L59. The `make dist` generates the source tarballs and the later Release step (not a workflow) makes it into a release. Also note the previous artifact upload step does just the opposite, it uploads the build artifacts to the CI artifacts system if it is *not* a tagged release.

Theoretically we build all tags, but the GH Releases system sometimes brings them to our attention sooner.
